crates/ty_python_semantic/src/types.rs

```diff
@@ -1271,6 +1338,99 @@ impl<'db> Type<'db> {
         }
     }
 
+    #[must_use]
+    pub(crate) fn recursive_type_normalized(
+        self,
+        db: &'db dyn Db,
+        visitor: &RecursiveTypeNormalizedVisitor<'db>,
+    ) -> Self {
+        if visitor.level() == 0 && self == visitor.div {
+            // int | Divergent = int | (int | (int | ...)) = int
+            return Type::Never;
+        } else if visitor.level() >= 1 && self.has_divergent_type(db, visitor.div) {
```

Comment:
If the level >= 1 and `self` does not contain `visitor.div`, do we need to keep walking `self` recursively, or could we just short-circuit and return `self`? It seems we will not change anything in the visit.

It's not really clear to me that we even need all the many added `recursive_type_normalized` methods in this PR, and all the `level` and `visit_no_shift` stuff either. Wouldn't it be an equivalent implementation if we just match on unions here, visit all their elements, and for any other type, check `has_divergent_type` and return `div` if so?